Correctly detect final path when extracting multiple tarballs

Flamefire opened the pull request:
The current implementation of `extract_file` detected folders/files from the first tarball when extracting the second.
Due to the definition of `find_base_dir` it will then return the parent path (usually `builddir`)
which is used as the `finalpath` for sources.
This leads issues requiring workarounds in e.g. the Bundle easyblock
where sources from all components are extracted into the same folder.

Fix by storing the old state of the target folder and detect of
extraction resulted in a single (top-level) folder.

As part of this work I enhanced the  documentation and test of find_base_dir and added tests for the expected and (previous) error cases of `extract_file`

`guess_start_dir` detects the fixed behavior already: If there is a `prefix/<startdir>/<startdir>` which doesn't exist `prefix/<startdir>` is used instead

E.g. GTK3 has start_dir set manually for components due to this bug which would now lead to builddir/foo/foo
IMO trivial to detect/fix

Remark: It is possible to have `find_base_dir` **NOT** change into the resulting directory which means we can do this explicitely when requested instead of undoing it. Not sure if anyone outside the main repo relies on that.


Followup: https://github.com/easybuilders/easybuild-easyblocks/pull/3961

Flamefire committed:
Enhance test_extract_file to verify returned path

Create an empty target folder so the extraction can change into the
extracted folder.

Flamefire committed:
Add test for `extract_file` of archive with multiple folders

Flamefire committed:
Enhance documentation and test of `find_base_dir`

Flamefire committed:
Correctly detect final path when extracting multiple tarballs

The current implementation of `extract_file` detected folders/files from the first tarball when extracting the second.
Due to the definition of `find_base_dir` it will then return the parent path (usually `builddir`)
which is used as the `finalpath` for sources.
This leads issues requiring workarounds in e.g. the Bundle easyblock
where sources from all components are extracted into the same folder.

Fix by storing the old state of the target folder and detect of
extraction resulted in a single (top-level) folder.

Flamefire committed:
Fix download of repo in *-from-pr

Adapt for `extract_dir` now returning the subfolder which it previously
did not because the downloaded tar file is in the same folder.

Flamefire committed:
Adapt test_apply_patch

boegel reviewed (changes requested):
@Flamefire Rather than adding (more) binary blobs to the repo, can we create `multi-0.0.tar.gz` and `multi-1.0.tar.gz` dynamically in the test itself?

That probably even helps with exposing what's being tested, since the structure within these tarballs is now not visible.

This request is fueled by [a recent example of binary blobs being abused](https://lwn.net/Articles/967180/).

I'm aware we have other binary blobs in here already, we should look into getting rid of those too.

Flamefire committed:
Remove uses for binaries from test_extract_file

Flamefire commented:
@boegel I replaced the added binaries by creating them in the test (replacing the commit to remove them from the history)

I also replaced all uses of other binaries in that test. The test-method I added should be easily usable for other tests.

What I noticed: Our `filetools.make_archive` is lacking a way to create an archive from a list of files or contents of a folder, i.e. exclude the parent folder from the hierarchy in the archive. See the new method I added.

Flamefire committed:
Test extracting just a file

Flamefire committed:
Handle extracting the same file to the same destination

When a source is added multiple times the content after the 2nd
extraction will be the same and no new files/folders are detected.
So `extract_file` would return the parent folder.
However if there was a clear result after the first extraction the same
result should be returned.

This works trivially until something else is extracted to the same
folder between the 2 extraction operations.

boegel commented:
@Flamefire merge conflict to resolve here...

Flamefire committed:
Fix typo

Flamefire committed:
Change into builddir when extracting multiple sources

For compatibility with the previous version only.
`find_base_dir` in `extract_file` couldn't find the base dir for
subsequent sources so fell back to the parent directory which is the
builddir in this case

Flamefire committed:
Change into directory of first source
